src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,792 @@
+toℕ∘2* (2suc  x) = cong (2 *ₙ_) (sym (ℕₚ.*-distribˡ-+ 2 1 (toℕ x)))
+
+toℕ∘suc : ∀ x → toℕ (suc x) ≡ 1+ (toℕ x)
+toℕ∘suc 0#        =  refl
+toℕ∘suc (2suc x)  =  cong (1+_ ∘ (2 *ₙ_)) (toℕ∘suc x)
+toℕ∘suc (suc2* x) =  ℕₚ.*-distribˡ-+ 2 1 (toℕ x)
+
+toℕ∘pred : ∀ x → toℕ (pred x) ≡ predₙ (toℕ x)
+toℕ∘pred 0#        =  refl
+toℕ∘pred (2suc x)  =  cong predₙ $ sym $ ℕₚ.*-distribˡ-+ 2 1 (toℕ x)
+toℕ∘pred (suc2* x) =  toℕ∘2* x
+
+toℕ∘fromℕ : toℕ ∘ fromℕ ≗ id
+toℕ∘fromℕ 0      = refl
+toℕ∘fromℕ (1+ n) = begin
+  toℕ (fromℕ (1+ n))   ≡⟨⟩
+  toℕ (suc (fromℕ n))  ≡⟨ toℕ∘suc (fromℕ n) ⟩
+  1+ (toℕ (fromℕ n))   ≡⟨ cong 1+_ (toℕ∘fromℕ n) ⟩
+  1+ n                 ∎
+
+toℕ-injective :  ∀ {x y} → toℕ x ≡ toℕ y → x ≡ y
+toℕ-injective {0#}     {0#}     _               =  refl
+toℕ-injective {2suc x} {2suc y} 2[1+xN]≡2[1+yN] =  cong 2suc x≡y
+  where
+  1+xN≡1+yN = ℕₚ.*-cancelˡ-≡ {1+ (toℕ x)} {1+ (toℕ y)} 1 2[1+xN]≡2[1+yN]
+  xN≡yN     = cong predₙ 1+xN≡1+yN
+  x≡y       = toℕ-injective xN≡yN
+toℕ-injective {2suc x} {suc2* y} 2[1+xN]≡1+2yN =
+  contradiction 2[1+xN]≡1+2yN (ℕₚ.even≢odd (1+ (toℕ x)) (toℕ y))
+toℕ-injective {suc2* x} {2suc y} 1+2xN≡2[1+yN] =
+  contradiction (sym 1+2xN≡2[1+yN]) (ℕₚ.even≢odd (1+ (toℕ y)) (toℕ x))
+toℕ-injective {suc2* x} {suc2* y} 1+2xN≡1+2yN =  cong suc2* x≡y
+  where
+  2xN≡2yN = cong predₙ 1+2xN≡1+2yN
+  xN≡yN = ℕₚ.*-cancelˡ-≡ 1 2xN≡2yN
+  x≡y = toℕ-injective xN≡yN
+
+toℕ-surjective : ∀ y → ∃ (λ x → toℕ x ≡ y)
+toℕ-surjective 0       =  (0# , refl)
+toℕ-surjective (1+ n)  with toℕ-surjective n
+... | (a , refl) = suc a , toℕ∘suc a
+
+fromℕ∘toℕ :  fromℕ ∘ toℕ ≗ id
+fromℕ∘toℕ =  toℕ-injective ∘ toℕ∘fromℕ ∘ toℕ
+
+fromℕ∘pred : ∀ n → fromℕ (predₙ n) ≡ pred (fromℕ n)
+fromℕ∘pred n = begin
+  fromℕ (predₙ n)        ≡⟨ cong (fromℕ ∘ predₙ) (sym (toℕ∘fromℕ n)) ⟩
+  fromℕ (predₙ (toℕ x))  ≡⟨ cong fromℕ (sym (toℕ∘pred x)) ⟩
+  fromℕ (toℕ (pred x))   ≡⟨ fromℕ∘toℕ (pred x) ⟩
+  pred x                 ≡⟨ refl ⟩
+  pred (fromℕ n)         ∎
+  where
+  x = fromℕ n
+
+x≡0⇒toℕ[x]≡0 :  ∀ {x} → x ≡ 0# → toℕ x ≡ 0
+x≡0⇒toℕ[x]≡0 {0#} _ = refl
+
+toℕ[x]≡0⇒x≡0 :  ∀ {x} → toℕ x ≡ 0 → x ≡ 0#
+toℕ[x]≡0⇒x≡0 {0#} _ = refl
+
+-----------------------------------------------------------------------
+-- Properties of _+_
+------------------------------------------------------------------------
+
+------------------------------------------------------------------------
+-- toℕ/fromℕ are homomorphisms for _+_
+
+toℕ-homo-+ :  ∀ x y → toℕ (x + y) ≡ toℕ x +ₙ toℕ y
+toℕ-homo-+ 0#        _        = refl
+toℕ-homo-+ (2suc  x) 0#       = cong 1+_ (sym (ℕₚ.+-identityʳ _))
+toℕ-homo-+ (suc2* x) 0#       = cong 1+_ (sym (ℕₚ.+-identityʳ _))
+toℕ-homo-+ (2suc x)  (2suc y) = begin
+  toℕ ((2suc x) + (2suc y))        ≡⟨⟩
+  toℕ (2suc (suc (x + y)))         ≡⟨⟩
+  2 *ₙ (1 +ₙ (toℕ (suc (x + y))))  ≡⟨ cong ((2 *ₙ_) ∘ 1+_) (toℕ∘suc (x + y)) ⟩
+  2 *ₙ (2 +ₙ toℕ (x + y))          ≡⟨ cong ((2 *ₙ_) ∘ (2 +ₙ_)) (toℕ-homo-+ x y) ⟩
+  2 *ₙ (2 +ₙ (m +ₙ n))             ≡⟨ cong ((2 *ₙ_) ∘ 1+_)
```

Comment:
As soon as you've pushed the `toℕ` all the way in, the solver from `Data.Nat.Solver` should be able to remove 4 or 5 lines from each case in the homomorphism proofs. For instance it should get you all the way down to the final line from here. I would check the other big proofs as well.

Reply:
Done